xbmc/pvr/providers/PVRProviders.cpp

```diff
@@ -0,0 +1,389 @@
+        provider->GetUniqueId() == newProvider->GetUniqueId())
+    {
+      found = true;
+      provider->UpdateEntry(newProvider, updateMode);
+    }
+  }
+
+  if (!found)
+  {
+    m_providers.emplace_back(newProvider);
+  }
+}
+
+std::vector<std::shared_ptr<CPVRProvider>> CPVRProvidersContainer::GetProvidersList() const
+{
+  CSingleLock lock(m_critSection);
+  return m_providers;
+}
+
+
+bool CPVRProviders::Load()
+{
+  // unload previous providers
+  Unload();
+
+  // load providers from database
+  bool bReturn = LoadFromDatabase();
+
+  // update from clients
+  Update();
+
+  return bReturn;
+}
+
+void CPVRProviders::Unload()
+{
+  // remove all tags
+  CSingleLock lock(m_critSection);
+  m_providers.clear();
+}
+
+bool CPVRProviders::LoadFromDatabase()
+{
+  const std::shared_ptr<CPVRDatabase> database = CServiceBroker::GetPVRManager().GetTVDatabase();
+  if (database)
+  {
+    bool bChanged = false;
+
+    CPVRProviders providers;
+    database->Get(providers);
+
+    for (auto& provider : providers.GetProvidersList())
+    {
+      bChanged |= (CheckAndAddEntry(provider, ProviderUpdateMode::BY_DATABASE) != nullptr);
+    }
+  }
+  return true;
+}
+
+bool CPVRProviders::Update()
+{
+  {
+    CSingleLock lock(m_critSection);
+    if (m_bIsUpdating)
+      return false;
+    m_bIsUpdating = true;
+  }
+
+  // Default client providers are the add-ons themselves, we retrieve both enabled
+  // and disbled add-ons as we don't want them removed from the DB
+  CPVRProviders newAddonProviderList;
+  std::vector<int> disabledClients;
+  std::vector<CVariant> clientProviderInfos =
+      CServiceBroker::GetPVRManager().Clients()->GetClientProviderInfos();
+  CLog::LogFC(LOGDEBUG, LOGPVR, "Adding default providers, found %d PVR add-ons",
+              clientProviderInfos.size());
+  for (const auto& clientInfo : clientProviderInfos)
+  {
+    auto addonProvider = std::make_shared<CPVRProvider>(
+        clientInfo["clientid"].asInteger(), clientInfo["name"].asString(),
+        clientInfo["icon"].asString(), clientInfo["thumb"].asString());
+
+    newAddonProviderList.CheckAndAddEntry(addonProvider, ProviderUpdateMode::BY_CLIENT);
+
+    if (!clientInfo["enabled"].asBoolean())
+      disabledClients.emplace_back(clientInfo["clientid"].asInteger());
+  }
+  UpdateDefaultEntries(newAddonProviderList);
+
+  // Client providers are retrieved from the clients
+  CLog::LogFC(LOGDEBUG, LOGPVR, "Updating providers");
+  CPVRProvidersContainer newProviderList;
+  std::vector<int> failedClients;
+  CServiceBroker::GetPVRManager().Clients()->GetProviders(&newProviderList, failedClients);
+  return UpdateClientEntries(newProviderList, failedClients, disabledClients);
+}
+
+bool CPVRProviders::UpdateDefaultEntries(const CPVRProvidersContainer& newProviders)
+{
+  bool bChanged = false;
+
+  CSingleLock lock(m_critSection);
+
+  /* go through the provider list and check for updated or new providers */
```

Comment:
I would go strictly with c++ comments for new source files. Here and everywhere else.

Reply:
Done